disasm-test/Main.hs

```diff
@@ -318,119 +365,240 @@ cube = TS.mkCUBE
-      return $ ss
-        <&> \s -> s { TS.expected = filter (rootExpSame s) $ TS.expected s }
+          addExpFilter s = s { TS.expected = filter (rootExpSame s) $ TS.expected s }
+      in fmap (fmap addExpFilter) . rangeMatch cb
   }
 
--- | A test failure.
-data TestFailure
-  = ParseError Error -- ^ A parser failure.
-    deriving (Typeable,Show)
 
-instance X.Exception TestFailure
+rangeMatch :: MonadIO m => TS.CUBE -> [TS.Sweets] -> m [TS.Sweets]
+rangeMatch cb swts = do
+  -- Perform ranged-matching of the llvm-range parameter against the
+  -- version of llvm (reported by llvm-as) to filter the tasty-sugar
+  -- expectations.
+  disasmOpts <- liftIO $ parseCmdLine
+  llvmver <- liftIO $ getLLVMAsVersion $ TO.lookupOption disasmOpts
+  TS.rangedParamAdjuster "llvm-range"
+    (readMaybe . drop (length ("pre-llvm" :: String)))
+    (<)
+    (vcVersioning llvmver ^? (_Right . major))
+    cb swts
+
+
+-- | Returns true if this particular test should be skipped, which is signalled
+-- by the expected file contents starting with "SKIP_TEST".  For test cases that
+-- require a minimum LLVM version, this technique is used to prevent running the
+-- test on older LLVM versions.
+skipTest :: TS.Expectation -> IO Bool
+skipTest expct =
+  ("SKIP_TEST" `L.isPrefixOf`) <$> L.readFile (TS.expectedFile expct)
+
+-- | There is a directory containing files which describe known bugs (one per
+-- file).  This function checks to see if the current test being defined
+-- corresponds to one of those known bugs, and if so, returns the name of the
+-- known bugs file and the summary string from that file.
+isKnownBug :: TS.Sweets -> IO (Maybe (FilePath, String))
+isKnownBug sweet = do
+  let kbdir = "disasm-test/known_bugs"
+  known <- listDirectory kbdir
+  let isSummaryLine = ("summary: " `isPrefixOf`)
+  let isMatchLine l = and [ "rootMatchName: " `isPrefixOf` l
+                          , TS.rootMatchName sweet `elem` words l
+                          ]
+  let checkMatch = \case
+        m@(Just _) -> const $ return m
+        Nothing -> \f -> do lse <- X.try (lines . toString <$> L.readFile (kbdir </> f))
+                            case lse of
+                              Left (_e :: IOError) ->
+                                -- just ignore this known_bugs file
+                                return Nothing
+                              Right ls ->
+                                let matches = not $ null $ filter isMatchLine ls
+                                in if matches
+                                   then let summary =
+                                              case find isSummaryLine ls of
+                                                Nothing -> "this is a known bug"
+                                                Just s -> s
+                                        in return $ Just (kbdir </> f, summary)
+                                   else return Nothing
+  foldM checkMatch Nothing known
```

Comment:
There's a fair bit of rightward drift in this code. It would be worth breaking out one or two helper functions to make this a bit more readable.

Reply:
I did a little more extensive refactoring of this (https://github.com/GaloisInc/llvm-pretty-bc-parser/pull/232/commits/428648630450356ee8fe30b1e9447a000550ca1e) and can re-address this if it went in a direction you have concerns about, but since it's in the original spirit of the PR, I'm going ahead with the merge.